add device agnostic tests & add helpful assertion for device agnostic tests

bgerstle opened the pull request:
Our app doesn't have a window when run as a test host, which results in `keyWindow` being `nil`.  This could also happen if your test target is a "Logic Test", and not an "Application Test."

```
$ ls -1 WikipediaUnitTests/ReferenceImages_64/NSAttributedString_WMFModifyTests
test0ReferenceOutput_iPhone_Simulator8_4_0x0@2x.png
testChangeGreenBackgroundColorsToBlue_iPhone_Simulator8_4_0x0@2x.png
testChangeReduceLineSpacing_iPhone_Simulator8_4_0x0@2x.png
testRemoveAllStrikeThroughs_iPhone_Simulator8_4_0x0@2x.png
testRemovingLink_iPhone_Simulator8_4_0x0@2x.png
```

I've added an assertion so others can get clearer feedback about what's going wrong and how to fix it.  FBSnapshotTestCase could arguably create a `keyWindow` implicitly, but I figured it's better to let users set up their test host properly instead of implicitly relying on the library to do it for them.

bgerstle commented:
looks like I need to download iOS 9 sim and re-record snapshots. downloading now... :hourglass_flowing_sand:

Grubas7 commented on FBSnapshotTestCase/FBSnapshotTestCasePlatform.m:
```diff
@@ -34,7 +34,7 @@ BOOL FBSnapshotTestCaseIs64Bit(void)
 NSString *FBDeviceAgnosticNormalizedFileName(NSString *fileName)
 {
   UIDevice *device = [UIDevice currentDevice];
-  CGSize screenSize = [[UIApplication sharedApplication] keyWindow].bounds.size;
```
I'm not sure if this change is OK.
UIWindow != UIScreen -> UIWindow is unique for application (especially useful for split view), and UIScreen is for whole device.

bgerstle replied:
That's true, but I figured using the screen would:

- Be equivalent to the window in most cases
- Work when the library is integrated as a framework (where `keyWindow` does not)

Maybe a more comprehensive way to solve this would be to generate the suffix based on the view being tested, which is probably what you really want anyway.  This makes even more sense when you consider that snapshots are generated based on the bounds of the specified view, even creating and "keying" a window if necessary.[0]

0: https://github.com/facebook/ios-snapshot-test-case/blob/master/FBSnapshotTestCase/Categories/UIImage+Snapshot.m#L46

nscoding replied:
hmm, we could check if the `keyWindow` is `nil` and if it is, create a window with the `[[UIScreen mainScreen] bounds]`

How does this sound?

bgerstle replied:
Actually, just had a realization: the problem doesn't have anything to do with frameworks.  Our app [short-circuits itself when being run as the test host](https://github.com/wikimedia/wikipedia-ios/blob/master/Wikipedia/Code/main.m#L10), which means no `AppDelegate`, which means no `keyWindow`.

So, this is more about handling our specific edge case, which means that a window should be made the `keyWindow` before suffixes are generated.  At which point, I'm wondering if we should be using a "dummy" app delegate (which just has a dummy window), or FBSnapshotTestCase should install one for me (e.g. if you were running a Logic Tests target which didn't even have a host).

Grubas7 replied:
I had similar problem:
I write some framework which is using UIApplication. Same framework doesn't have UIApplication, so I created TestApp, which hosts tests.
So check if this resolve your problem:
1. Add new Application target
2. Choose your framework test target
3. Go to `general` tab
4. Set your new Application target as `Host Application`
5. Make sure that your Application target contains framework tests in test schema

nscoding replied:
nice, if that works maybe we can assert and explain how to solve this

bgerstle replied:
@Grubas7 the short-circuit is intentional, and our app is already the "Host" for our tests.  This removes the unnecessary "link in both targets" step when adding new app code.

My question to you guys is, should FBSnapshotTestCase be defensive by setting the `keyWindow` itself (which it already tries to do), and if so, should it guarantee that `keyWindow

bgerstle replied:
My question about determines whether the solution should be _either_:

- Add an assertion and explain what users should do (set a host app or otherwise create a key window when visual tests are running)
- Create a `keyWindow` manually if necessary

Grubas7 commented on FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m:
```diff
@@ -0,0 +1,38 @@
+- (void)tearDown {
+    // Put teardown code here. This method is called after the invocation of each test method in the class.
+    [super tearDown];
+}
```
Empty method. Should be delated.

bgerstle replied:
:ok_hand:

bgerstle commented:
@Grubas7 @nscoding I reverted my change using `UIScreen` and added an assertion instead.  This should prompt users (like me) who don't have a key window to go ahead and create one.

Grubas7 commented on FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m:
```diff
@@ -0,0 +1,34 @@
+/*
```
Are those test still valid?

Grubas7 commented:
@bgerstle Cool. I wonder there should be some info about this in README. @nscoding ?

nscoding commented on FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m:
```diff
@@ -0,0 +1,34 @@
+{
+    UIView *redView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
+    redView.backgroundColor = [UIColor redColor];
+    FBSnapshotVerifyView(redView, nil);
```
Travis is hitting the assertion that you added, why?

bgerstle replied:
Good question, will have to investigate..
On Sun, Dec 13, 2015 at 17:30 Patrick Chamelo <notifications@github.com>
wrote:

> In
> FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m
> <https://github.com/facebook/ios-snapshot-test-case/pull/137#discussion_r47451523>
> :
>
> > +@end
> > +
> > +@implementation FBSnapshotTestCaseDeviceAgnosticTests
> > +
> > +- (void)setUp
> > +{
> > +    [super setUp];
> > +    self.deviceAgnostic = YES;
> > +    self.recordMode = NO;
> > +}
> > +
> > +- (void)testViewSnapshot
> > +{
> > +    UIView *redView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
> > +    redView.backgroundColor = [UIColor redColor];
> > +    FBSnapshotVerifyView(redView, nil);
>
> Travis is hitting the assertion that you added, why?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ios-snapshot-test-case/pull/137/files#r47451523>
> .
>

bgerstle replied:
Because the `keyWindow` is `nil`... :confounded: Even though the app delegate is certainly [making it key and visible](https://github.com/facebook/ios-snapshot-test-case/blob/master/FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/FBAppDelegate.m#L22). Any ideas?

bgerstle replied:
Also @Grubas7 this is why the tests were added

bgerstle replied:
Should I:

- Somehow wait for the application to have a key window, maybe using `expectationWithPredicate`?
- Fall back to the application's delegate's window?

nscoding replied:
hmm I just cloned your PR, [UIApplication sharedApplication] is nil at this point. I believe mainly because of the project doesn't have a host application.

![screen shot 2015-12-16 at 11 23 43 am](https://cloud.githubusercontent.com/assets/185995/11851423/7b583e44-a3e7-11e5-85d6-b1dfc7226b60.png)

nscoding replied:
Either we fix the project to have a host application (I am leaning toward this) or we don't rely on `[UIApplication sharedApplication]` but I don't see a clean way to get the window.

Thoughts?

bgerstle replied:
Oh wow, didn't even think to check that... When I ran the tests locally I put a break point in `FBAppDelegate` and saw it was being called.  I'll fix that straightaway.

bgerstle replied:
@nscoding were you checking the demo project or the framework?  The demo project has a test host, which corroborates what I saw when debugging:

![image](https://cloud.githubusercontent.com/assets/444217/11851895/21310916-a403-11e5-889c-dc4404c4520c.png)

As the device agnostic tests I added are in the demo project, I don't think changing the framework test target to have a host will make a difference here.

nscoding replied:
oh, I was look at the project itself, you are right, this is configured correctly but yet `[UIApplication sharedApplication]` is still `nil`

bgerstle replied:
what do you suggest? poll `keyWIndow` in `FBSnapshotTestCase.setUp`? e.g.:

``` objc
- (void)setUp 
{
  // other stuff
  [self expectationWithPredicate:[NSPredicate predicateWithBlock:^ (...) {
    return evaluatedObject.keyWindow != nil;
  } ...];
}
```

nscoding replied:
Sorry, I had to look at it a bit to figure out what is going on and there are a couple of issues:

1. The build script that I added on travis is building and testing both [iPhone 5 and iPhone 6](https://github.com/facebook/ios-snapshot-test-case/blob/master/build.sh#L25-L26), so we have to record the image for both of these devices.  

2. The [build script](https://github.com/facebook/ios-snapshot-test-case/blob/master/build.sh#L7-L11) runs the FBSnapshotControllerTests and in particular [testFailedImageWithDeviceAgnosticShouldHaveModelOnName](https://github.com/facebook/ios-snapshot-test-case/blob/master/FBSnapshotTestCaseTests/FBSnapshotControllerTests.m#L69). These tests don't have an application associated with, so we hit the assertion. Maybe we can make the window a parameter and simply write tests to make sure that FBDeviceAgnosticNormalizedFileName is working as expected.

nscoding replied:
@bgerstle thoughts on my latest comment?

nscoding replied:
oh I didn't see the updated commits

bgerstle replied:
@nscoding yeah, sorry I didn't respond, but better late than never:

> The build script that I added on travis is building and testing both iPhone 5 and iPhone 6, so we have to record the image for both of these devices.

I already figured that out when Travis failed and downloaded/recorded the proper screenshots.

> testFailedImageWithDeviceAgnosticShouldHaveModelOnName. These tests don't have an application associated with, so we hit the assertion. 

Those tests are in the Demo project, which does have a host application (see above screenshot).  I _think_ the problem was that `UIImage+Snapshot` was causing issues when installing a "dummy" window when given a view without a window.  This somehow caused the `keyWindow` to return `nil` when device-agnostic tests were run—which I determined after seeing each test pass in isolation.  As a result, I've modified `UIImage+Snapshot` to use the application's `keyWindow` if the given view doesn't have a window (or if it isn't a window itself).  Let me know if you think this is acceptable.

> Maybe we can make the window a parameter and simply write tests to make sure that FBDeviceAgnosticNormalizedFileName is working as expected.

That could be useful for verifying the formatting of test names (to prevent unintended regressions and thus break tests for users), but doesn't address the problems targeted by this PR.

bgerstle commented:
@Grubas7 @nscoding this should be ready to go, see my [latest comment](https://github.com/facebook/ios-snapshot-test-case/pull/137#discussion_r48477202) for a brief explanation on the approach and c9a9530 for the implementation

Grubas7 commented on FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m:
```diff
@@ -0,0 +1,24 @@
+
+@implementation UIApplication (StrictKeyWindow)
+
+- (UIWindow*)fb_strictKeyWindow {
```
Missing space before asterisk.

bgerstle replied:
Have you guys considered using uncrustify or clang-format? 😅 While I'm at it, should I remove the comment headers (Created by Brian Gerstle etc.)?

bgerstle replied:
went ahead and did it anyway

Grubas7 commented on FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m:
```diff
@@ -0,0 +1,26 @@
+@implementation UIApplication (StrictKeyWindow)
+
+- (UIWindow *)fb_strictKeyWindow {
+  UIWindow *keyWindow = [[UIApplication sharedApplication] keyWindow];
```
`keyWindow` is a property, so please use dot notation.

Grubas7 commented on FBSnapshotTestCase/Categories/UIImage+Snapshot.m:
```diff
@@ -57,6 +62,11 @@ + (UIImage *)fb_imageForView:(UIView *)view
   UIImage *snapshot = UIGraphicsGetImageFromCurrentImageContext();
   UIGraphicsEndImageContext();
+
+  if (removeFromSuperview) {
```
`if ([window.subviews containObject:view])` IMO opinion is less error prone.

bgerstle replied:
I disagree, since there's an edge case where the condition you're testing is actually setup by the user, and therefore would be incorrect for us to tear down. The intent here is to only tear down what we ourselves setup on line 50/55.

Grubas7 replied:
OK, you right.

Grubas7 commented on FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m:
```diff
@@ -0,0 +1,34 @@
+    self.recordMode = NO;
+}
+
+- (void)testViewSnapshot
```
Test name too general.

Grubas7 commented on FBSnapshotTestCaseTests/FBSnapshotControllerTests.m:
```diff
@@ -66,24 +66,6 @@ - (void)testCompareReferenceImageWithVeryLowToleranceShouldMatch
     XCTAssertTrue([controller compareReferenceImage:referenceImage toImage:testImage tolerance:.001 error:nil]);
 }
 
-- (void)testFailedImageWithDeviceAgnosticShouldHaveModelOnName
```
Why did you remove this test?

Grubas7 commented on FBSnapshotTestCase/FBSnapshotTestCasePlatform.m:
```diff
@@ -34,7 +35,8 @@ BOOL FBSnapshotTestCaseIs64Bit(void)
 {
   UIDevice *device = [UIDevice currentDevice];
-  CGSize screenSize = [[UIApplication sharedApplication] keyWindow].bounds.size;
+  UIWindow* keyWindow = [[UIApplication sharedApplication] fb_strictKeyWindow];
```
Pointer alignment ;)

Grubas7 commented:
Looks good :+1:

bgerstle commented:
@Grubas7 thanks for the feedback! all your comments should be addressed :ok_hand:

Grubas7 commented:
@bgerstle could you squash your commits?

bgerstle commented:
@Grubas7 squashed & amended commit message

nscoding commented on FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m:
```diff
@@ -0,0 +1,26 @@
+
+@implementation UIApplication (StrictKeyWindow)
+
+- (UIWindow *)fb_strictKeyWindow {
```
new line before {

nscoding commented on FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.h:
```diff
@@ -0,0 +1,20 @@
+@interface UIApplication (StrictKeyWindow)
+
+/**
+  @return The recdeiver's @c keyWindow. Raises an assertion if @c nil.
```
recdeiver's  typo

nscoding commented on FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m:
```diff
@@ -0,0 +1,34 @@
+    self.recordMode = NO;
+}
+
+- (void)testSuccessfulComparisonOfDeviceAgnosticReferenceImages
```
can we also run this test (and record the image) for an iPhone 6?

bgerstle replied:
Interestingly, it's already run on both 5 & 6 iPhone models, but the application itself isn't filling the screen:

![image](https://cloud.githubusercontent.com/assets/444217/12079129/e09a2660-b1fb-11e5-9091-cb2ff438d7ca.png)

I'll see if I can fix this quickly.  Good catch that this should be verified for different screen sizes.

nscoding commented:
Thank you for working on this, it looks great, I left a couple of comments (nit picks really).

bgerstle commented:
@nscoding addressed all of your comments, including the iPhone 6 screen size issue (and re-recorded reference image)

nscoding commented:
testSuccessfulComparisonOfDeviceAgnosticReferenceImages tests are failing, it's looking for 320x568 (iPhone 5 right? but we didn't provide this one)

bgerstle commented:
Ahh that's different now too, as it used to be 320x480, forgot to re-record
those
On Sun, Jan 3, 2016 at 18:20 Patrick Chamelo <notifications@github.com>
wrote:

> testSuccessfulComparisonOfDeviceAgnosticReferenceImages tests are failing,
> it's looking for 320x568 (iPhone 5 right? but we didn't provide this one)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ios-snapshot-test-case/pull/137#issuecomment-168551718>
> .
>

bgerstle commented:
@nscoding should be good now :checkered_flag:

nscoding commented on FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Launch Screen.storyboard:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="9531" systemVersion="15C50" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" initialViewController="01J-lp-oVM">
```
Do we need this?

bgerstle replied:
Yes, the launch screen fixes the issue where the app doesn't fill the
screen.

On Fri, Jan 8, 2016 at 12:50 Patrick Chamelo <notifications@github.com>
wrote:

> In FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Launch Screen.storyboard
> <https://github.com/facebook/ios-snapshot-test-case/pull/137#discussion_r49236579>
> :
>
> > @@ -0,0 +1,50 @@
> > +<?xml version="1.0" encoding="UTF-8" standalone="no"?>
> > +<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="9531" systemVersion="15C50" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" initialViewController="01J-lp-oVM">
>
> Do we need this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ios-snapshot-test-case/pull/137/files#r49236579>
> .
>

nscoding commented on FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Images.xcassets/LaunchImage.launchimage/Contents.json:
```diff
@@ -4,15 +4,134 @@
       "orientation" : "portrait",
       "idiom" : "iphone",
       "extent" : "full-screen",
+      "minimum-system-version" : "8.0",
```
I guess this file was introduced for the device size right?

bgerstle replied:
Yeah, one of those things that changed just by touching it. Didn't think it
was worth messing with git to revert it.

On Fri, Jan 8, 2016 at 12:51 Patrick Chamelo <notifications@github.com>
wrote:

> In
> FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Images.xcassets/LaunchImage.launchimage/Contents.json
> <https://github.com/facebook/ios-snapshot-test-case/pull/137#discussion_r49236613>
> :
>
> > @@ -4,15 +4,134 @@
> >        "orientation" : "portrait",
> >        "idiom" : "iphone",
> >        "extent" : "full-screen",
> > +      "minimum-system-version" : "8.0",
>
> I guess this file was introduced for the device size right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ios-snapshot-test-case/pull/137/files#r49236613>
> .
>
